File: src/test/java/cc/vileda/kalfor/CombineHandlerTest.java
package cc.vileda.kalfor;

import io.restassured.RestAssured;
import io.vertx.core.json.JsonArray;
import io.vertx.core.json.JsonObject;
import io.vertx.ext.unit.Async;
import io.vertx.ext.unit.TestContext;
import io.vertx.ext.unit.junit.VertxUnitRunner;
import io.vertx.rxjava.core.RxHelper;
import io.vertx.rxjava.core.Vertx;
import io.vertx.rxjava.core.http.HttpClient;
import org.junit.After;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import rx.Observable;

import java.io.IOException;
import java.net.ServerSocket;
import java.util.Arrays;

import static io.restassured.RestAssured.get;
import static io.restassured.RestAssured.given;
import static org.hamcrest.Matchers.containsString;


@RunWith(VertxUnitRunner.class)
public class CombineHandlerTest
{
		private final int kalforPort = getRandomPort();
		private final int remotePort = getRandomPort();

		public CombineHandlerTest() throws IOException {}

		@Before
		public void setUp(TestContext context) throws IOException
		{
				final Vertx vertx = Vertx.vertx();
				final Async async = context.async();
				final Observable<String> deployVerticle = RxHelper.deployVerticle(vertx, new RestApiMock(remotePort));
				final Observable<String> deployVerticle2 = RxHelper.deployVerticle(
						vertx,
						new KalforVerticle(new KalforOptions(new Endpoint("http://localhost:" + remotePort), kalforPort))
				);
				deployVerticle
						.subscribe(s -> {
								System.out.println(s);
								deployVerticle2.subscribe(s1 -> {
										System.out.println(s1);
										async.complete();
								});
						});
		}

		@Test
		public void restApiMockShouldRespond()
		{
				get("http://localhost:"+remotePort+"/test")
						.then()
						.body(containsString(new JsonObject().put("foo", "bar").encodePrettily()));
		}

		@Test
		public void combineHandlerShouldCombine()
		{
				final String given = new JsonArray(Arrays.asList(
						new JsonObject().put("firstKey", "/test"),
						new JsonObject().put("secondKey", "/test")
				)).encodePrettily();
				final String expected = new JsonObject()
						.put("firstKey", new JsonObject().put("foo", "bar"))
						.put("secondKey", new JsonObject().put("foo", "bar"))
						.encodePrettily();
				given()
						.body(given)
						.when()
							.post("http://localhost:"+kalforPort+"/combine")
						.then()
							.body(containsString(expected));
		}

		private static int getRandomPort() throws IOException
		{
				ServerSocket socket = new ServerSocket(0);
				int port = socket.getLocalPort();
				socket.close();

				return port;
		}
}

File: library/src/test/kotlin/cc/vileda/kalfor/handler/CombineHandlerTest.kt
package cc.vileda.kalfor.handler

import com.jayway.restassured.RestAssured
import com.jayway.restassured.RestAssured.get
import com.jayway.restassured.response.Header
import io.vertx.core.json.JsonArray
import io.vertx.core.json.JsonObject
import io.vertx.ext.unit.TestContext
import io.vertx.ext.unit.junit.VertxUnitRunner
import io.vertx.rxjava.core.RxHelper
import io.vertx.rxjava.core.Vertx
import org.apache.http.HttpStatus
import org.apache.http.entity.ContentType
import org.hamcrest.CoreMatchers.containsString
import org.hamcrest.Matchers
import org.junit.BeforeClass
import org.junit.Ignore
import org.junit.Test
import org.junit.runner.RunWith
import java.net.ServerSocket
import java.net.URLEncoder
import java.util.*

@RunWith(VertxUnitRunner::class)
class CombineHandlerTest {
    @Test
    fun staticFileServerMockShouldRespond() {
        get("http://localhost:$staticFileServerMockPort/test1.css")
                .then()
                .body(containsString("#test1 { margin: 0 auto; }"))

        get("http://localhost:$staticFileServerMockPort/test2.css")
                .then()
                .body(containsString("#test2 { margin: 0 auto; }"))
    }

    @Test
    fun restApiMockShouldRespond() {
        get("http://localhost:$restApiMockPort/test")
                .then()
                .body(containsString(JsonObject().put("1foo", "1bar").encodePrettily()))
    }

    @Test
    fun combineHandlerShouldCombineViaPost() {
        val given = JsonArray(listOf(KalforRequest(
                "http://localhost:$restApiMockPort",
                listOf(KalforProxyHeader("x-foo", "bar")),
                Arrays.asList(
                        KalforProxyRequest("1firstKey", "/test"),
                        KalforProxyRequest("2secondKey", "/test2"))))).encodePrettily()

        val expected = JsonObject()
                .put("1firstKey", JsonObject().put("1foo", "1bar"))
                .put("2secondKey", JsonObject().put("2foo", "2bar"))
                .encodePrettily()

        RestAssured
                .given()
                .body(given)
                .header(Header("Content-Type", "application/json"))
                .`when`()
                .post("http://localhost:$kalforPort/combine")
                .then()
                .body(Matchers.containsString(expected))
    }

    @Test
    fun combineHandlerShouldCombineMultipleViaPost() {
        val req1 = KalforRequest(
                "http://localhost:$restApiMockPort",
                listOf(KalforProxyHeader("x-foo", "bar")),
                Arrays.asList(
                        KalforProxyRequest("1firstKey", "/test"),
                        KalforProxyRequest("2secondKey", "/test2")))

        val req2 = KalforRequest(
                "http://localhost:$restApi2MockPort",
                listOf(KalforProxyHeader("x-foo", "bar")),
                Arrays.asList(
                        KalforProxyRequest("3thirdKey", "/test3"),
                        KalforProxyRequest("4fourthKey", "/test4")))

        val given = JsonArray(listOf(req1, req2)).encodePrettily()

        val expected = JsonObject()
                .put("1firstKey", JsonObject().put("1foo", "1bar"))
                .put("2secondKey", JsonObject().put("2foo", "2bar"))
                .put("3thirdKey", JsonObject().put("1foo", "1bar"))
                .put("4fourthKey", JsonObject().put("2foo", "2bar").put("a3foo", "a3bar"))
                .encodePrettily()

        RestAssured
                .given()
                .body(given)
                .header(Header("Content-Type", "application/json"))
                .`when`()
                .post("http://localhost:$kalforPort/combine")
                .then()
                .body(Matchers.containsString(expected))
    }

    @Test
    fun combineHandlerShouldCombineViaGet() {
        val given = listOf(
                "http://localhost:$staticFileServerMockPort/test1.css",
                "http://localhost:$staticFileServerMockPort/test2.css"
        ).map { URLEncoder.encode(it, Charsets.UTF_8.name()) }

        val expected = """#test1 { margin: 0 auto; }
#test2 { margin: 0 auto; }"""

        RestAssured
                .given()
                .param("c", given)
                .header("x-foo", "bar")
                .header("content-type", ContentType.TEXT_PLAIN)
                .`when`()
                .get("http://localhost:$kalforPort/combine")
                .then()
                .body(Matchers.containsString(expected))
    }

    @Test
    fun combineHandlerShouldRespondWithBadRequestOnInvalidRequest() {
        val given = JsonArray(Arrays.asList(
                JsonObject().put("foo", "bar"),
                JsonObject().put("baz", JsonObject().put("bar", "foo")))).encodePrettily()

        RestAssured
                .given()
                .body(given)
                .header(Header("Content-Type", "application/json"))
                .`when`()
                .post("http://localhost:$kalforPort/combine")
                .then()
                .statusCode(Matchers.`is`(HttpStatus.SC_BAD_REQUEST))
                .body(Matchers.containsString("{\n" +
                        "  \"error\" : [ \"object has missing required properties " +
                        "([\\\"proxyBaseUrl\\\",\\\"proxyRequests\\\"])\", " +
                        "\"object has missing required " +
                        "properties ([\\\"proxyBaseUrl\\\",\\\"proxyRequests\\\"])\" ]\n" +
                        "}"))
    }

    companion object {
        private val kalforPort = randomPort
        private val restApiMockPort = randomPort
        private val restApi2MockPort = randomPort
        private val staticFileServerMockPort = randomPort

        private val randomPort: Int
            get() {
                val socket = ServerSocket(0)
                val port = socket.localPort
                socket.close()

                return port
            }

        @BeforeClass
        @JvmStatic
        fun setUp(context: TestContext) {
            val vertx = Vertx.vertx()
            val async = context.async()
            val restApiMockVerticle = RxHelper.deployVerticle(vertx, RestApiMock(restApiMockPort))
            val restApi2MockVerticle = RxHelper.deployVerticle(vertx, RestApi2Mock(restApi2MockPort))
            val staticFileServerMockVerticle = RxHelper.deployVerticle(vertx, StaticFileServerMock(staticFileServerMockPort))
            val kalforVerticle = RxHelper.deployVerticle(vertx, KalforTestVerticle(kalforPort))
    
            restApiMockVerticle
                    .mergeWith(restApi2MockVerticle)
                    .mergeWith(staticFileServerMockVerticle)
                    .mergeWith(kalforVerticle)
                    .subscribe {
                        println("deployed Kalfor " + it)
                        async.complete()
                    }
        }
    }
}